L67-cpp-callback-api.md

```diff
@@ -0,0 +1,126 @@
+   - `Finish`: for server RPCs to provide completion status. This initiates the asynchronous transmission of the status to the client. Must be used for all server RPCs, including those that have already been cancelled. *Server only*.
+   - _Constraints_: There may not be more than one concurrent `StartRead` and more than one concurrent `StartWrite` in-progress for a given RPC. However, `StartRead` and `StartWrite` may be concurrent. These operations are considered done when the `OnReadDone` or `OnWriteDone` reactions is invoked. There must be exactly one call of `Finish` at the server and `StartCall` at the client. There may be no more than one use of `StartWritesDone` at the client or `StartSendInitialMetadata` at the server. Any invocation of `StartSendInitialMetadata` at the server must happen before an invocation of `StartWrite` or `Finish`.
+1. Operation-completion reaction methods: notification of completion of asynchronous RPC activity. These are all virtual methods that default to an empty function (i.e., `{}`) but may be overridden by the application's reactor definition.
+   - `OnReadDone`, `OnWriteDone`, `OnWritesDoneDone`: methods invoked by the library at the completion of an explicitly-specified streaming RPC operation. Provides a `bool ok` argument to indicate whether the completion was successful. *Client and server*. 
+   - `OnReadInitialMetadataDone`: notifying client RPCs that the server has sent initial metadata. It has a bool `ok` argument to indicate whether the completion was successful. This is automatically invoked by the library for all RPCs; it is uncommonly used however. *Client only*
+   - `OnCancel`: for server RPCs only, called if an RPC is canceled before it has a chance to successfully send status to the client side. Should be used for any cleanup associated with cancellation. Note that even cancelled RPCs _must_ get a `Finish` call to cleanup all their library state. *Server only*
+   - `OnDone`: called when all outstanding RPC operations are completed for both client and server RPCs. For the client-side, it additionally provides the status that the server sent with its `Finish` call. Should be used for any application-level RPC-specific cleanup. *Client and server.*
+   - _Thread safety_: the above calls may take place concurrently, except that `OnDone` will always take place after all other reactions. No further RPC operations should be issued after `OnDone` is invoked.
+   - **IMPORTANT USAGE NOTE** : code in any reaction must not block for an arbitrary amount of time since reactions are executed on a finite-sized library-controlled threadpool. If any long-term blocking operations (like sleeps, file I/O, synchronous RPCs, or waiting on a condition variable) must be invoked as part of the application logic, then it is important to push that outside the reaction so that the reaction can complete in a timely fashion. One way of doing this is to push that code to a separate application-controlled thread.
```

Comment:
```suggestion
   - **IMPORTANT USAGE NOTE** : code in any reaction must not block for an arbitrary amount of time since reactions are executed on a finite-sized library-controlled threadpool. If any long-term blocking operations (like sleeps, file I/O, synchronous RPCs, or waiting on a condition variable) must be invoked as part of the application logic, then it is important to push that outside the reaction so that the reaction can complete in a timely fashion. One way of doing this is to push that code to a separate application-controlled work queue.
```

Consider using "work queue" instead of "thread" to minimize the change for misinterpretation that the recommendation is that start a new thread for each blocking call. A "work queue" may do that, but it may also be able to handle multiple blocking calls with one thread.

Reply:
Let me think about this one; I think I want something stronger than work queue even if thread isn't the right term since work queue isn't really a well-defined term whereas thread is.